prometheus_alerting_consistency.md

```diff
@@ -0,0 +1,105 @@
+  labels:
+    severity: critical
+  annotations:
+    summary: 'The backend service is stuck in a non-ready state'
+    description: 'The backend service {{ $labels.name }} in the {{ $labels.namespace }} namespace, managed by operator {{ $labels.pod }} has been in a non-ready state for 10 minutes'
+    sop_url: 'https://example.com/backend_service_stuck.asciidoc'
+```
+
+This alert fires if a backend service has *not* been ready for the last 10 minutes.
+This is a clear example of a critical issue that represents a threat to the operability of the service, and likely warrants paging someone.
+The alert has a clear summary and description annotations, and it links to a SOP with information on investigating and resolving the issue.
+
+The group of critical alerts should be small, very well-defined, highly documented, polished and with a high bar set for entry.
+
+## Warning Alerts
+
+The vast majority of alerts should use this severity.
+Issues at the warning level should be addressed in a timely manner, but don't pose an immediate threat to the operation of the service as a whole.
+
+Timeline: ~60 minutes
+
+If your alert does not meet the criteria in "Critical Alerts" above, it belongs to the warning level or lower.
+
+Use warning level alerts for reporting conditions that may lead to inability to deliver individual features of the service, but not the service as a whole.
+Most alerts are likely to be warnings.
+Configure warning level alerts so that they do not fire until components have sufficient time to try to recover from the interruption automatically.
+You should have a policy for triaging warning alerts and responding in a timely manner e.g. issues created automatically on a backlog that is being actively worked from.
+
+Example warning alert:
+
+```yaml
+- alert: ZookeeperPersistentVolumeFillingUp
+  expr: (kubelet_volume_stats_available_bytes{persistentvolumeclaim=~"data-(.+)-zookeeper-[0-9]+"} / kubelet_volume_stats_capacity_bytes{persistentvolumeclaim=~"data-(.+)-zookeeper-[0-9]+"} < 0.15) and predict_linear(kubelet_volume_stats_available_bytes{persistentvolumeclaim=~"data-(.+)-zookeeper-[0-9]+"}[6h], 4 * 24 * 3600) < 0
+  for: 1h
+  labels:
+    severity: warning
+  annotations:
+    summary: 'Zookeeper PersistentVolume is filling up.'
+    description: 'Based on recent sampling, the Zookeeper PersistentVolume claimed by {{ $labels.persistentvolumeclaim }} in Namespace {{ $labels.namespace }} is expected to fill up within four days. Currently {{ $value | humanizePercentage }} is available.'
+    sop_url: 'https://example.com/persistent_volume_filling.asciidoc'
+```
+
+This alert fires if one or more Zookeeper volumes are getting close to filling up.
+The alert has a clear name and informative summary and description annotations, and it links to a SOP with information on investigating and resolving the issue.
+The timeline is appropriate for allowing the service to resolve the issue itself, avoiding the need to alert SRE.
```

Comment:
hmm, `for: 1h` seems like a good way to keep from bothering humans for short-term effects.  But I'm having a bit of trouble wrapping my head around how it interacts with the 6h range being used for [`predict_linear`][1].  Might be worth checking with the monitoring folks to see if there are recommendations like "when using `predict_linear`, you don't need to bother with `for`" or "when using `predict_linear`, set `for` to something larger than the range given to `predict_linear`.

[1]: https://prometheus.io/docs/prometheus/latest/querying/functions/#predict_linear

Reply:
I think both time ranges (one for `for`, one for `predict_linear`) serve different purposes and work well in concert.

1. `predict_linear`: This time range tells prometheus which samples to select to run a simple linear regression. It will work if there are [at least 2 samples present](https://github.com/prometheus/prometheus/blob/fc3f1b8ded810a1263f72d37db9db4e8ce726120/promql/functions.go#L774-L788). So in this query, the `predict_linear` part says, "given the last 6 hours of data, the volume will fill up in 4 days or less".
2. The `for` clause keeps this from firing unless the query as been true for an hour. This normalizes big spikes in the samples considered.

An example where this matters is when you have a big spike in temporary data written, i.e. data that has a short life time. If the `for` clause is left out, a big enough spike will trigger the alert, no matter how many samples are passed to `predict_linear` (or the time range is so large that it'll never trigger). What this misses though is the when the data is deleted shortly after, the alert will already have fired.
With the `for` clause there is a chance to catch this sort of scenario.
I don't think we can give a general rule how these work together best, it depends on the signal we want an alert on.

For this example in particular I think the question is: Are we expecting spikes in temporary data? If not it might make sense to drop the `for` clause.

Additional thought:

- Maybe drop the first part of the query. This alert will only fire if `predict_linear` falls under zero _and_ the volume has only 15% space available. The last part is not mentioned in the alert name or description. I would separate that into a second alert to improve clarity.
- Be more explicit in the alert description. Why not something like `Based on the last 6 hours worth of data, the Zookeeper PersistentVolume claimed by {{ $labels.persistentvolumeclaim }} in Namespace {{ $labels.namespace }} is expected to fill up within four days or less.`

Reply:
> If the `for` clause is left out, a big enough spike will trigger the alert, no matter how many samples are passed to `predict_linear`...
> ...
> For this example in particular I think the question is: Are we expecting spikes in temporary data? If not it might make sense to drop the `for` clause.

Fair, so possibly no general rule.

> Maybe drop the first part of the query. This alert will only fire if `predict_linear` falls under zero and the volume has only 15% space available.

You need the <15% bit to turn off the alert quickly after space is freed, because it might take a while after cleanup before the 6h linear regression pulls back up into a positive extrapolated prediction.

Some extreme cases for disk usage:

* Most false-happy prediction would be 0% free for 6h, a single scrape at 100% free, and then right back to 0% free.  It would take 3h after the return to 0% before the predicted future dropped back below 0%.
* Most false-sad prediction would be trending straight from 100% free to 10% free over 6h, and then settles in around 10% without dropping further.  As long as you didn't start dropping again, you'd be safe with 10% free, but the 4d prediction would be negative for several hours until the earlier very-free numbers started to drop out of the 6h `predict_linear`.

The `for: 1h` doesn't make much difference in either of those.  I guess you could have an oscillation with a <1h period, where the peak-free was over 15%, or the period was right around 6h and predicted zero time dropped under the 4h threshold, but only for <1h at a time.  I guess what I'm deciding is that the thresholds here need to be tied fairly tightly to guesses about likely behavior (e.g. "our garbage-collection routine kicks in at <15% free, and usually runs for <15m"), and selecting these values with a more generic understanding of the system being monitored is pretty risky.

Reply:
> You need the <15% bit to turn off the alert quickly after space is freed, because it might take a while after cleanup before the 6h linear regression pulls back up into a positive extrapolated prediction.

Good point yeah. My main thought was that it would be nice to have a low space alert as well. If that is of interest, the filling-up alert could query the synthetic ALERTS metric and replace the first part with something like `ALERTS{alertname="ZookeeperPersistentVolumeAvailableLow",alertstate="firing", persistentvolumeclaim="foo"} = 1`

I agree that more insight into the IO behavior will help to write a more precise alert. Even then some weird IO pattern can probably make this noisy. That's why I think making the description as explicit as possible is good practice. The more information the respondent has, the better they can structure their response. 
"Based on the IO rates of the last 6 hours, you PVC will be full in $time" is much more actionable then "Your PVC is will be full soon", to exaggerate a bit.

Reply:
@jan--f @wking 
This conversation is excellent, and I think it would be great to capture what you think the outcome of it is.
As an fyi, I've since pushed a simpler promql expression for the example alert so as to keep the focus of the document on consistency of alerts.
However, I feel like there's some wording updates needed to capture what you're saying here in a general way,
as well as any tips for using predict_linear you think are relevant

Reply:
> As an fyi, I've since pushed a simpler promql expression for the example alert so as to keep the focus of the document on consistency of alerts.

In this case I would remove all references to `predict_linear` and I would also change the example since alerting on a fixed threshold value is something I wouldn't recommend. https://github.com/openshift/cluster-monitoring-operator/blob/bc45b875b3aa4c4f36abf35b532e3bc9294b75b6/assets/prometheus-k8s/prometheus-rule.yaml#L114-L124 might be a good alternative (though it is Prometheus specific, it should be easy to grasp).

Reply:
Thanks for the suggestion @simonpasquier 
I've updated the example & text.